src/main/python/dataproducts/consumption/consumption_metrics_v2.py

```diff
@@ -0,0 +1,574 @@
+    ).select(
+        func.col("dimensions.sid"),
+        func.col("dimensions.pdata.id").alias("pdata_id"),
+        func.col("dimensions.type"),
+        func.col("dimensions.mode"),
+        func.col("dimensions.did"),
+        func.col("object.id").alias("object_id"),
+        func.col("edata.eks.time_spent"),
+        func.col("object.rollup.l1")
+    )
+    app = data.filter(
+        func.col('type').isin('app') &
+        func.col('pdata_id').isin(config['context']['pdata']['id']['app'], config['context']['pdata']['id']['desktop'])
+    )
+    app_df = app.groupBy(
+        func.col('pdata_id')
+    ).agg(
+        func.count('sid').alias('Total App Sessions'),
+        func.countDistinct('did').alias('Total Devices on App'),
+        (func.sum('time_spent') / 3600).alias('Total Time on App (in hours)')
+    ).toPandas()
+    app_df['x_index'] = 0
+    app_df.set_index("x_index", inplace=True)
+    x_app = app_df.pivot(columns='pdata_id')
+    result_loc_.joinpath(date_.strftime('%Y-%m-%d')).mkdir(exist_ok=True)
+    x_app.to_csv(result_loc_.joinpath(date_.strftime('%Y-%m-%d'), 'app_sessions.csv'), index=False)
+    post_data_to_blob(result_loc_.joinpath(date_.strftime('%Y-%m-%d'), 'app_sessions.csv'), backup=True)
+    play = data.filter(func.col('mode').isin('play'))
+    content = spark.read.csv(
+        str(result_loc_.parent.joinpath('tb_metadata', date_.strftime('%Y-%m-%d'), 'textbook_snapshot.csv')),
+        header=True
+    ).select(
+        func.col('identifier'),
+        func.col('channel')
+    ).distinct()
+    play_df = play.join(
+        content,
+        play.l1 == content.identifier,
+        how='left'
+    ).groupBy(
+        func.col('channel'),
+        func.col('pdata_id')
+    ).agg(
+        func.count('sid').alias('Total Content Plays'),
+        func.countDistinct('did').alias('Total Devices that played content'),
+        (func.sum('time_spent') / 3600).alias('Content Play Time (in hours)')
+    ).toPandas()
+    x_play = play_df.pivot(index='channel', columns='pdata_id')
+    x_play.to_csv(result_loc_.joinpath(date_.strftime('%Y-%m-%d'), 'plays.csv'))
+    post_data_to_blob(result_loc_.joinpath(date_.strftime('%Y-%m-%d'), 'plays.csv'), backup=True)
+    spark.stop()
+
+
+def dialscans(result_loc_, date_):
+    """
+    compute failed/successful scans by channel
+    :param result_loc_: pathlib.Path object to store resultant CSV at.
+    :param date_: datetime object to use in query and path
+    :return: None
+    """
+    spark = SparkSession.builder.appName("DialcodeScans").master("local[*]").getOrCreate()
+    account_name = os.environ['AZURE_STORAGE_ACCOUNT']
+    account_key = os.environ['AZURE_STORAGE_ACCESS_KEY']
+    container = 'telemetry-data-store'
+    spark.conf.set('fs.azure.account.key.{}.blob.core.windows.net'.format(account_name), account_key)
+    path = 'wasbs://{}@{}.blob.core.windows.net/telemetry-denormalized/raw/{}-*'.format(container, account_name,
```

Comment:
Code is duplicated and also the data is loaded twice. The denorm data needs to be read once and use it compute all metrics. Do not read the data multiple times